File: java/src/automation-root/automation-core/src/main/java/com/automation/core/components/FluentBase.java
package com.automation.core.components;

import com.automation.core.logging.Logger;

import java.lang.reflect.InvocationTargetException;

@SuppressWarnings("WeakerAccess")
public abstract class FluentBase implements Fluent {
    private final Logger logger;

    public FluentBase(Logger logger) {
        this.logger = logger;
    }

    public Logger getLogger() {
        return logger;
    }

    @Override
    public <T> T changeContext(Class c)
            throws InvocationTargetException, NoSuchMethodException, InstantiationException, IllegalAccessException {
        // factory
        return generateObject(c, null);
    }

    @Override
    public <T> T changeContext(Class c, Logger logger)
            throws InvocationTargetException, NoSuchMethodException, InstantiationException, IllegalAccessException {
        // factory
        return generateObject(c, logger);
    }

    @Override
    public abstract <T> T changeContext(Class c, String application, Logger logger)
            throws ClassNotFoundException, InvocationTargetException, NoSuchMethodException, InstantiationException, IllegalAccessException;

    @Override
    public abstract <T> T changeContext(Class c, String application)
            throws ClassNotFoundException, InvocationTargetException, NoSuchMethodException, InstantiationException, IllegalAccessException;

    @Override
    public abstract <T> T changeContext(String type, String application)
            throws ClassNotFoundException, InvocationTargetException, NoSuchMethodException, InstantiationException, IllegalAccessException;

    abstract <T> T generateObject(Class c, Logger logger)
            throws NoSuchMethodException, IllegalAccessException, InvocationTargetException, InstantiationException;
}

File: csharp/src/AutomationRoot/Automation.Core/Components/FluentBase.cs
using Automation.Core.Logging;
using System;

namespace Automation.Core.Components
{
    public abstract class FluentBase : IFluent
    {
        protected FluentBase(ILogger logger)
        {
            Logger = logger;
        }

        public ILogger Logger { get; }

        public T ChangeContext<T>()
        {
            var instance = Create<T>(null, null);
            Logger.Debug($"instance of [{GetType()?.FullName}] created");
            return instance;
        }

        public T ChangeContext<T>(ILogger logger)
        {
            return Create<T>(null, logger);
        }

        public abstract T ChangeContext<T>(string application, ILogger logger);

        public abstract T ChangeContext<T>(string application);

        public abstract T ChangeContext<T>(string type, string application);

        internal abstract T Create<T>(Type type, ILogger logger);
    }
}